### PR TITLE
Fake container exec/logs support for in-process docker CRI integration

### PR DESCRIPTION
**Pull request by yujuhong**
This is necessary to unblock other work on docker integration, while we are addressing
`logs` and `exec` in the meantime.

This is part of #31459 and #33189

/cc @kubernetes/sig-node 

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/33301)

### CODE REVIEW

**Comment by yujuhong** [comment]
I am running the node e2e test locally to see how this affects the test results. If this works out well, perhaps we can add a presubmit builder that's only triggered by keywords so that we can test related PRs.

@yifan-gu, the first commit moves the `ConvertPodStatusToRunningPod` you modified last time to `pkg/kubelet/container` so that we don't end up having import cycles. 
I think in general, we should try avoid importing kuberuntime in existing implementation (`dockertools`, `rkt`). The opposite is also true, but since it's WIP, we may sometimes have to borrow some code from existing implementation as an interim solution. 

/cc @Random-Liu  as we've talked about this yesterday.

**Comment by yujuhong** [comment]
With this PR, node e2e went from `FAIL! -- 27 Passed | 75 Failed | 0 Pending | 37 Skipped` to `112 Passed | 24 Failed | 0 Pending | 3 Skipped`. 

Ready for reviews.

**Comment by Random-Liu** [comment]
> With this PR, node e2e went from FAIL! -- 27 Passed | 75 Failed | 0 Pending | 37 Skipped to 112 Passed | 24 Failed | 0 Pending | 3 Skipped.

@yujuhong I think you are running all the tests. :) In fact that includes Serial and Flaky test. I believe some tests are failed because you were running them in parallel. If so there should be more test pass. :)

**Comment by yujuhong** [comment]
Ah..I see. Didn't even notice that the numbers don't add up. Now I'm a bit suspicious as I know the shim lacks a lot of features.

**Comment by Random-Liu** [comment]
@yujuhong What command are you using to run the test?

**Comment by yifan-gu** [comment]
LGTM.

File: pkg/kubelet/dockershim/legacy.go [hunk 1]
```diff
@@ -42,3 +42,13 @@ func (ds *dockerService) GetContainerLogs(pod *api.Pod, containerID kubecontaine
 func (ds *dockerService) PortForward(pod *kubecontainer.Pod, port uint16, stream io.ReadWriteCloser) error {
 	return fmt.Errorf("not implemented")
 }
+
+func (ds *dockerService) ExecInContainer(containerID kubecontainer.ContainerID, cmd []string, stdin io.Reader, stdout, stderr io.WriteCloser, tty bool, resize <-chan term.Size) error {
+	container, err := ds.client.InspectContainer(containerID.ID)
+	if err != nil {
+		return err
+	}
+
```
Comment (Random-Liu) [inline pkg/kubelet/dockershim/legacy.go]: Still add the running check even though this is only a temporary solution.

File: pkg/kubelet/kuberuntime/kuberuntime_container.go [hunk 2]
```diff
@@ -521,6 +529,13 @@ func (m *kubeGenericRuntimeManager) GetContainerLogs(pod *api.Pod, containerID k
 // tty.
 // TODO: handle terminal resizing, refer https://github.com/kubernetes/kubernetes/issues/29579
 func (m *kubeGenericRuntimeManager) ExecInContainer(containerID kubecontainer.ContainerID, cmd []string, stdin io.Reader, stdout, stderr io.WriteCloser, tty bool, resize <-chan term.Size) error {
+	// Use `docker exec` directly for in-process docker integration for
+	// now to unblock other tests.
+	// TODO: remove this hack after exec is defined in CRI.
+	if ds, ok := m.runtimeService.(dockershim.DockerLegacyService); ok {
+		return ds.ExecInContainer(containerID, cmd, stdin, stdout, stderr, tty, resize)
+	}
+
```
Comment (Random-Liu) [inline pkg/kubelet/kuberuntime/kuberuntime_container.go]: Extra space comparing with function above.

**Review by Random-Liu — COMMENTED** [review]
LGTM with nits

File: pkg/kubelet/dockershim/legacy.go (hunk 1 above)
Comment (yujuhong) [inline pkg/kubelet/dockershim/legacy.go]: Didn't think it was necessary since docker will error out. Will add it.

**Comment by yujuhong** [comment]
Addressed the nits.

**Comment by feiskyer** [comment]
> 112 Passed | 24 Failed | 0 Pending | 3 Skipped.

Great. And some of them is because lacking features, e.g. init containers.

**Pull request merged**